server/tournament/models.py

```diff
@@ -0,0 +1,314 @@
+
+    title = models.CharField(max_length=255)
+    start_date = models.DateField()
+    end_date = models.DateField()
+    team_registration_start_date = models.DateField()
+    team_registration_end_date = models.DateField()
+    player_registration_start_date = models.DateField()
+    player_registration_end_date = models.DateField()
+    ultimate_central_id = models.PositiveIntegerField(unique=True, null=True, blank=True)
+    ultimate_central_slug = models.SlugField(max_length=200, default="unknown")
+    location = models.CharField(max_length=255, default="unknown")
+    type = models.CharField(max_length=3, choices=Type.choices, default=Type.MIXED)
+    slug = models.SlugField(null=True, blank=True, db_index=True)
+    max_num_teams = models.PositiveSmallIntegerField(blank=True, null=True)
+    series = models.ForeignKey(Series, on_delete=models.SET_NULL, blank=True, null=True)
+
+    def save(self, *args: Any, **kwargs: Any) -> None:
+        if not self.slug:
+            slug = self.get_slug()
+            self.slug = slug
+            self.ultimate_central_slug = slug
+        return super().save(*args, **kwargs)
+
+    def get_slug(self) -> str:
+        slug = slugify_max(self.title, 45)
+        unique_slug = slug
+
+        number = 1
+        while Event.objects.filter(slug=unique_slug).exists():
+            unique_slug = f"{unique_slug}-{number}"
+            number += 1
+
+        return unique_slug
+
+
+class Tournament(ExportModelOperationsMixin("tournament"), models.Model):  # type: ignore[misc]
+    class Status(models.TextChoices):
+        DRAFT = "DFT", _("Draft")
+        REGISTERING = "REG", _("Registering")
+        SCHEDULING = "SCH", _("Scheduling")
+        LIVE = "LIV", _("Live")
+        COMPLETED = "COM", _("Completed")
+
+    event = models.OneToOneField(Event, on_delete=models.CASCADE, unique=True)
+    teams = models.ManyToManyField(Team, related_name="tournaments")
+    status = models.CharField(max_length=3, choices=Status.choices, default=Status.DRAFT)
+    logo_light = models.FileField(upload_to="tournament_logos/", blank=True, max_length=256)
+    logo_dark = models.FileField(upload_to="tournament_logos/", blank=True, max_length=256)
+    rules = models.TextField(blank=True, null=True)
+
+    initial_seeding = models.JSONField(default=dict)
+    current_seeding = models.JSONField(default=dict)
+    spirit_ranking = models.JSONField(default=list)
+
+    use_uc_registrations = models.BooleanField(default=False)
+
+
+@receiver(m2m_changed, sender=Tournament.teams.through)
+def update_seeding_on_teams_change(
+    sender: Any, instance: Tournament, action: str, **kwargs: Any
+) -> None:
+    if action == "post_add":
+        seeding = {}
+
+        for i, team in enumerate(instance.teams.all(), start=1):
+            seeding[i] = team.id
+
+        instance.initial_seeding = seeding
+        instance.current_seeding = seeding
+        instance.save()
+
+    if action == "post_remove":
```

Comment:
minor comment: Why added this seperately? Can just combine in the top to have if action is post_add or post_remove do the same thing?